support run work agent outside of managed cluster

zhujian7 opened the pull request:
Signed-off-by: zhujian <jiazhu@redhat.com>

related issue: https://github.com/open-cluster-management-io/registration-operator/issues/158

---
todo:
- [ ] add description and script to deploy work agent in Hosted mode. (not contained in this PR)

qiujian16 commented on pkg/spoke/spokeagent.go:
```diff
@@ -43,6 +46,7 @@ func (o *WorkloadAgentOptions) AddFlags(cmd *cobra.Command) {
 	flags := cmd.Flags()
 	// This command only supports reading from config
 	flags.StringVar(&o.HubKubeconfigFile, "hub-kubeconfig", o.HubKubeconfigFile, "Location of kubeconfig file to connect to hub cluster.")
+	flags.StringVar(&o.SpokeKubeconfigFile, "spoke-kubeconfig", o.SpokeKubeconfigFile, "Location of kubeconfig file to connect to spoke cluster.")
```
can't we use the built-in --kubeconfig flag here?

zhujian7 replied:
I  think it is better not, the built-in `--kubeconfig` in our case represents the cluster where the work-agent pod run.

qiujian16 replied:
this should be explicit callout that if this is not set, but --kubeconfig is set. kubeconfig will be used to connect to spoke.

zhujian7 committed:
support run work agent outside of managed cluster

add a flag spoke-kubeconfig to specify the kubeconfig of the apoke/managed cluster

Signed-off-by: zhujian <jiazhu@redhat.com>

qiujian16 commented on pkg/spoke/spokeagent.go:
```diff
@@ -24,10 +26,12 @@ import (
-	Burst             int
+	HubKubeconfigFile   string
+	SpokeKubeconfigFile string
+	SpokeKubeconfig     *rest.Config
```
why do we need this?

zhujian7 replied:
This was originally used for integration test, but now I think it can be deleted

qiujian16 commented on pkg/spoke/spokeagent.go:
```diff
@@ -64,28 +69,39 @@ func (o *WorkloadAgentOptions) RunWorkloadAgent(ctx context.Context, controllerC
-	spokeDynamicClient, err := dynamic.NewForConfig(spokeRestConfig)
+	// load spoke client config and create spoke clients,
+	// the work agent may running not in the spoke/managed cluster.
+	if o.SpokeKubeconfig == nil {
```
I do not see how it can be not nil

qiujian16 commented on pkg/spoke/spokeagent.go:
```diff
@@ -64,28 +69,39 @@ func (o *WorkloadAgentOptions) RunWorkloadAgent(ctx context.Context, controllerC
+		if o.SpokeKubeconfigFile != "" {
+			o.SpokeKubeconfig, err = clientcmd.BuildConfigFromFlags("" /* leave masterurl as empty */, o.SpokeKubeconfigFile)
+			if err != nil {
+				return fmt.Errorf("unable to load spoke kubeconfig from file %q: %w", o.SpokeKubeconfigFile, err)
```
I would make it a func

```go
 func (o *WorkloadAgentOptions) spokeKubeConfig(controllerContext) *rest.Config, error {
    if o.SpokeKubeconfigFile != "" {
        return clientcmd.BuildConfigFromFlags("" /* leave masterurl as empty */, o.SpokeKubeconfigFile)
    }

   return controllerContext.KubeConfig
}
```

zhujian7 committed:
delete field SpokeKubeconfig

Signed-off-by: zhujian <jiazhu@redhat.com>

zhujian7 commented:
/retest

qiujian16 reviewed (commented):
/approve
/lgtm
